### PR TITLE
Add power management mode selection

### LINKED ISSUE #132: Power mode option for good battery life
## Feature

I work on a pretty beefy MSI Workstation laptop, but I don't want the animations from Elementary, for example, when I want all the battery life I can get. This is a feature that has existed on Windows OS since (around) Windows 7. They also have other power options like "High performance" that disables lots of internal OS features to improve performance but I'm not sure that's necessary with Elementary because it is mostly slim.

This feature would be best implemented as a on/off button on the panel for the case of batteries. It should disable animations in the operating system and any other very minute details that could otherwise be prevented to save on battery.

**Alternatives considered**

I have turned my screen brightness down to the very limit I'm comfortable with and my keyboard brightness is completely off. Works great, but more could be done.

**Additional context**

image

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/89038119-power-mode-option-for-good-battery-life?utm_campaign=plugin&utm_content=tracker%2F60466177&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60466177&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by SubhadeepJasu**
Fixes #132

This Pull Requests adds support for power management mode in the WingPanel power indicator.

Screenshots:
Light Mode
Dark Mode

### CODE REVIEW

**Comment by SubhadeepJasu** [comment]
There is a section of code where given access to the "settings-daemon" schema it syncs the power mode with that instead of syncing with power-profiles interface directly. Just need to be sure that's okay, I can't test it as that settings schema doesn't exist on my system and I am on OS7.

File: .vscode/settings.json
Comment (danirabbit) [inline .vscode/settings.json:1]: Oops probably don't want to add your vscode settings here

File: src/Widgets/PopoverWidget.vala
Comment (danirabbit) [inline src/Widgets/PopoverWidget.vala:149]: Oh this is an interesting syntax I hadn't seen before. Is this a built-in null check? very cool!

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:6]: We don't use `using` statements in elementary code style

File: src/Widgets/PopoverWidget.vala
Comment (danirabbit) [inline src/Widgets/PopoverWidget.vala:78]: This argument seems weird. Why are we initializing this class with false?

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:28]: The Object call here should really only contain things passed in as construction arguments. So it would only contain `on_battery`, but it seems `on_battery` is always false? so it's not clear why this is a construction argument at all

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:26]: Spacing in a box is `0` by default so no need to specify

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:25]: You can use only the actual name of the constant for a property that takes a namespaced constant like 
```suggestion
            orientation: VERTICAL,
```

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:52]: `construct` is already your kind of `build_ui` function. it's the function that gets called once when the class is initialized, so we don't need to make a separate `build_ui` function

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:57]: Instead of adding icons from resource like this, you can just add them to the icon theme and then reference them by name. So in the Indicator class you could add:
`
Gtk.IconTheme.get_default ().add_resource_path ("/io/elementary/desktop/wingpanel/power");
`

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:63]: padding and margins should be in multiples of 3

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:66]: Should probably just use `add` here. `pack_start` and `pack_end` don't exist for `Box` in GTK4

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:111]:
```suggestion
                    add (saver_radio);
                    break;
                case "balanced":
                    add (balanced_radio);
                    break;
                case "performance":
                    add (performance_radio);
```

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:156]: leftover debugging code?

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:9]: probably should default to false?

File: data/io.elementary.desktop.wingpanel.power.gschema.xml
Comment (danirabbit) [inline data/io.elementary.desktop.wingpanel.power.gschema.xml:14]: This appears unused?

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:61]: Maybe this should be `3`? It looks like there's twice as much space between the icons and labels here as there is for connected devices

**Review by danirabbit — CHANGES_REQUESTED** [review]
This appears to work for me on OS 8 daily with no crashes! Great job scaling the icons they look great.

I left some comments about code style etc. Also don't forget to add new classes to POTFILES!

Also on desktop I'm missing the separator between power profiles and the settings menuitem:

Screenshot from 2024-08-12 10 38 41

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:9]: It starts with true, until we encounter an error during construction of the object and then its set to false.

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:25]: That's not working 😕

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:28]: It's a placehoder code, if it access the settings-daemon schema then it chooses the appropriate key to save/load the settings from. This is same as how it's there in the power switchboard plug.

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:52]: I usually separate the ui building code and the event handling code in two separate functions. I can put it all in one place if you say, it's just my style.

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:61]: I did this to ensure the icon appears visually midway between the radio circle and the label.

File: src/Widgets/PopoverWidget.vala
Comment (SubhadeepJasu) [reply src/Widgets/PopoverWidget.vala:78]: This is place holder code taken from power switchboard plug. This allows usnig a specific settings key based on whether the laptop is on battery or not.

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:6]: Fixed

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala]: Fixed

File: .vscode/settings.json
Comment (SubhadeepJasu) [reply .vscode/settings.json]: Oops. Fixed

File: src/Widgets/PopoverWidget.vala
Comment (SubhadeepJasu) [reply src/Widgets/PopoverWidget.vala:149]: Yes this is a null-safe expression. This will only run the function if `power_mode_list` is not null. I learnt it in C# and found out Vala supports it too.

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala]: Yup Fixed

**Comment by lenemter** [comment]
@SubhadeepJasu Translations are updated automatically on push into `master`. Please revert the latest commit :)

**Comment by SubhadeepJasu** [comment]
> @SubhadeepJasu Translations are updated automatically on push into `master`. Please revert the latest commit :)

Ok

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [reply src/Widgets/PowerModeList.vala:8]: Seems like this could cause a future race where this property claims the class is initialized before its actually finished/ready

Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:8]: Wouldn't the check happen after the class has been constructed and allocated? The function in question is get_proxy_sync which is a blocking code. Or am I misunderstanding something?

File: src/Widgets/PopoverWidget.vala
Comment (danirabbit) [reply src/Widgets/PopoverWidget.vala:78]: Yeah we should make sure to clean up any placeholder stuff before merging please

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [reply src/Widgets/PowerModeList.vala:8]: Hm yeah I get what your'e saying. But if someone refactors it in the future to be async. I dunno. I'm not gonna block you on it, but it doesn't seem good to me to like speculate optimistically about future state instead of making sure that we plan to fail in the correct state

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [reply src/Widgets/PowerModeList.vala:24]: Ah maybe it doesn't work in Object calls, sorry. Either way, previous comment about only having construction properties in the object call still stands! If a new constructor is added in the future for example these should really be in the `construct` block

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:21]: If this is always initialized `false` I think this really shouldn't be here. A future commit can re-add this property if it's actually used in the future

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [reply src/Widgets/PowerModeList.vala:55]: I'm not gonna block on it, but we usually don't do this

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:60]: I would really prefer this to be in the `Indicator` class since we probably will resource the battery icons for the display widget in the future so we don't want to duplicate adding the resource path

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:68]: In GTK4 there's no margin property, so new code should prepare for that:
```suggestion
            margin_top = 3,
            margin_end = 3,
            margin_bottom = 3,
            margin_start = 3
```

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:75]: Can we use `child` property instead here? There's no `add` for CheckButton in GTK4 so saves us having to port that in the future

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:88]: Can we use `group` property instead of `from_widget` constructor? There's no `from_widget` in GTK4

File: src/Widgets/PowerModeList.vala
Comment (danirabbit) [inline src/Widgets/PowerModeList.vala:128]: It looks like setting from the settings key does update the backend already, so this should probably only touch the backend if settings is null right?

**Review by danirabbit — CHANGES_REQUESTED** [review]
Thanks for quick turnaround! Left a few more comments about getting ahead of GTK4 changes, and some other follow up. Looking much better!

File: src/Widgets/PopoverWidget.vala
Comment (SubhadeepJasu) [reply src/Widgets/PopoverWidget.vala:78]: Binding it to dm.has_battery to see if it works.

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:60]: Done

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:68]: Done

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:75]: Done

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:128]: Fixed

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:21]: I have binded this to actual plugged in status.

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala:24]: Done

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala]: Done!

File: src/Widgets/PowerModeList.vala
Comment (SubhadeepJasu) [reply src/Widgets/PowerModeList.vala]: Implemented it, so no longer place holder.

**Review by danirabbit — APPROVED** [review]
Looks good, works great! Thanks so much!

**Comment by SubhadeepJasu** [comment]
> Looks good, works great! Thanks so much!

Thank you too!

**Pull request merged**
